docs/internal/3i.md

```diff
@@ -0,0 +1,169 @@
+
+3i eliminates these constraints by introducing a user-space routing layer between cages, grates, and the underlying microvisor. Calls can be dispatched directly between grates or delegated to the microvisor when necessary, achieving kernel-level extensibility while keeping all new logic external to the kernel’s TCB.
+
+### High-level Concepts
+
+In traditional operating systems, a process makes a system call which traps into the kernel.  Every process which makes a system call ends up trapping into the same kernel routine.  In essence, there is one system call table which is shared by every process.
+
+In contrast, in Lind, 3i provides a per-cage system call table.  A cage can write a function that is a system call handler (or handler, for short) and then register a handler ( register_handler ) for a specific system call of a cage (security will be discussed later).  Whenever that cage makes a system call ( make_syscall ), it ends up calling that function.  Since every cage has its own system call table, the handlers for different cages may vary.
+
+As a term of convenience, we call a cage which processes system calls a "grate".  This is meant to convey the mental model of a cage calling down towards the microvisor / kernel and having a grate filter or transform or handle those system calls.  However, a grate is simply a cage and there is no special handling code or permission for it in 3i or the rest of the system.  A grate may tend to make different system calls from a normal application, but it is still a cage, much like strace is still a normal Linux process that happens to use system calls like ptrace which are otherwise rare.
+
+One important feature needed by a grate is the ability to read and write the memory of a cage which makes a system call it intercepts.  For example, to handle a write system call, the grate must be able to read data out of the calling cage's buffer, which involves reading the calling cage's memory.  3i provides a function ( copy_data_between_cages ) to enable this feature safely.  
+
+Let's consider a special case for a moment, that of a grate that wants to count the number of times the write system call is called.  This grate wants to increment a counter and then make the cage's write system call using the grate's own system call table.  If the cage is writing large amounts of data, copying this information would be extremely wasteful given that the grate does not care about its contents or need to read it.  Hence, make_syscall enables different arguments to a syscall to come from different cages.  This means that the grate can execute the write system call and indicate that the pointer to the write buffer is in the cage's address space.  This ability to specify arguments that live in different address spaces is needed across a variety of grates and so 3i's make_syscall function supports specifying a cageid per argument.
```

Comment:
This example is a bit unclear. Count the number of calls to `write()` in what scope? A single cage, all cages running within the lifetime of a Lind process? Answering this will also help explain why arguments from different cages are needed in this example.